vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/helper/AuthenticationRequestParser.kt

```diff
@@ -0,0 +1,93 @@
+import at.asitplus.wallet.lib.oidc.AuthenticationRequestParameters
+import at.asitplus.wallet.lib.oidc.AuthenticationRequestParametersFrom
+import at.asitplus.wallet.lib.oidc.AuthenticationResponseParameters
+import at.asitplus.wallet.lib.oidc.AuthenticationResponseResult
+import at.asitplus.wallet.lib.oidc.OpenIdConstants
+import at.asitplus.wallet.lib.oidc.OpenIdConstants.Errors
+import at.asitplus.wallet.lib.oidc.RemoteResourceRetrieverFunction
+import at.asitplus.wallet.lib.oidc.RequestObjectJwsVerifier
+import at.asitplus.wallet.lib.oidvci.OAuth2Exception
+import at.asitplus.wallet.lib.oidvci.decodeFromUrlQuery
+import io.github.aakira.napier.Napier
+import io.ktor.http.Url
+import io.ktor.util.flattenEntries
+
+internal class AuthenticationRequestParser(
+    /**
+     * Need to implement if resources are defined by reference, i.e. the URL for a [JsonWebKeySet],
+     * or the authentication request itself as `request_uri`, or `presentation_definition_uri`.
+     * Implementations need to fetch the url passed in, and return either the body, if there is one,
+     * or the HTTP header `Location`, i.e. if the server sends the request object as a redirect.
+     */
+    private val remoteResourceRetriever: RemoteResourceRetrieverFunction,
+    /**
+     * Need to verify the request object serialized as a JWS,
+     * which may be signed with a pre-registered key (see [OpenIdConstants.ClientIdScheme.PRE_REGISTERED]).
+     */
+    private val requestObjectJwsVerifier: RequestObjectJwsVerifier,
+) {
+    companion object {
+        fun createWithDefaults(
+            remoteResourceRetriever: RemoteResourceRetrieverFunction? = null,
+            requestObjectJwsVerifier: RequestObjectJwsVerifier? = null,
+        ) = AuthenticationRequestParser(
+            remoteResourceRetriever = remoteResourceRetriever ?: { null },
+            requestObjectJwsVerifier = requestObjectJwsVerifier ?: RequestObjectJwsVerifier { _, _ -> true },
+        )
+    }
+
+    /**
+     * Pass in the URL sent by the Verifier (containing the [AuthenticationRequestParameters] as query parameters),
+     * to create [AuthenticationResponseParameters] that can be sent back to the Verifier, see
+     * [AuthenticationResponseResult].
+     */
+    suspend fun parseAuthenticationRequestParameters(input: String): KmmResult<AuthenticationRequestParametersFrom> = catching {
+        // maybe it is a request JWS
+        val parsedParams = kotlin.run { parseRequestObjectJws(input) }
+            ?: kotlin.runCatching { // maybe it's in the URL parameters
+                Url(input).let {
+                    val params = it.parameters.flattenEntries().toMap()
+                        .decodeFromUrlQuery<AuthenticationRequestParameters>()
+                    AuthenticationRequestParametersFrom.Uri(it, params)
+                }
+            }.onFailure { it.printStackTrace() }.getOrNull()
+            ?: kotlin.runCatching {  // maybe it is already a JSON string
```

Comment:
`catching` all the way down?

Reply:
What exactly do you mean by that?

Reply:
If the comment is to be interpreted like in https://github.com/a-sit-plus/kmm-vc-library/pull/95#discussion_r1668509349 then yes, this is indeed intentional because the error is handeled properly